tests/ECC/Ed25519.hs

```diff
@@ -0,0 +1,119 @@
+            propertyHold [ eqTest "distributive left"  ((sa `Ed25519.scalarMul` sb) `Ed25519.scalarAdd` (sa `Ed25519.scalarMul` sc))
+                                                       (sa `Ed25519.scalarMul` (sb `Ed25519.scalarAdd` sc))
+                         , eqTest "distributive right" ((sb `Ed25519.scalarMul` sa) `Ed25519.scalarAdd` (sc `Ed25519.scalarMul` sa))
+                                                       ((sb `Ed25519.scalarAdd` sc) `Ed25519.scalarMul` sa)
+                         ]
+        ]
+    , testGroup "point arithmetic"
+        [ testProperty "pointDecode.pointEncode==id" $ \p ->
+            let bs = Ed25519.pointEncode p :: ByteString
+                p' = Ed25519.pointDecode bs
+             in CryptoPassed p `propertyEq` p'
+        , testProperty "pointEncode.pointDecode==id" $ \p ->
+            let b  = Ed25519.pointEncode p :: ByteString
+                p' = Ed25519.pointDecode b
+                b' = Ed25519.pointEncode `fmap` p'
+             in CryptoPassed b `propertyEq` b'
+        , testProperty "addition with identity" $ \p ->
+            propertyHold [ eqTest "identity left"  p (Ed25519.pointAdd p0 p)
+                         , eqTest "identity right" p (Ed25519.pointAdd p p0)
+                         ]
+        , testProperty "addition associative" $ \pa pb pc ->
+            Ed25519.pointAdd pa (Ed25519.pointAdd pb pc) === Ed25519.pointAdd (Ed25519.pointAdd pa pb) pc
+        , testProperty "addition commutative" $ \pa pb ->
+            Ed25519.pointAdd pa pb === Ed25519.pointAdd pb pa
+        , testProperty "negation" $ \p ->
+            p0 `propertyEq` Ed25519.pointAdd p (Ed25519.pointNegate p)
+        , testProperty "doubling" $ \p ->
+            Ed25519.pointAdd p p `propertyEq` Ed25519.pointDouble p
+        , testProperty "scalarmult with zero" $ \p ->
+            p0 `propertyEq` Ed25519.pointMul s0 p
+        , testProperty "scalarmult with one" $ \p ->
+            p `propertyEq` Ed25519.pointMul s1 p
+        , testProperty "scalarmult with two" $ \p ->
+            Ed25519.pointDouble p `propertyEq` Ed25519.pointMul s2 p
+        , testProperty "scalarmult with curve order - 1" $ \p ->
+            Ed25519.pointNegate p `propertyEq` Ed25519.pointMul sI p
+        , testProperty "scalarmult commutative" $ \a b ->
+            Ed25519.pointMul a (Ed25519.toPoint b) === Ed25519.pointMul b (Ed25519.toPoint a)
+        , testProperty "scalarmult distributive" $ \x y p ->
+            let pR = Ed25519.pointMul x p `Ed25519.pointAdd` Ed25519.pointMul y p
+             in pR `propertyEq` Ed25519.pointMul (x `Ed25519.scalarAdd` y) p
+        , testProperty "double scalarmult" $ \n1 n2 p ->
+            let pR = Ed25519.pointAdd (Ed25519.toPoint n1) (Ed25519.pointMul n2 p)
+             in pR `propertyEq` Ed25519.pointsMulVarTime n1 n2 p
+        ]
+    ]
+  where
+    p0 = Ed25519.toPoint s0
+    CryptoPassed s0 = Ed25519.scalarDecodeLong ("" :: ByteString)
```

Comment:
it might be more informative to replace a pattern match by a `throwCryptoError`. although I suspect we'ld be losing the location error right now by doing so, so might be adding a CallStack context to `throwCryptoError` which would be an improvement by itself anyway ?

Reply:
I'll look into it.